### PR TITLE
feat: third purity value

### LINKED ISSUE #7013: Third Purity value
Instead of being pure or impure, it should be pure, impure, or control-pure

### LINKED ISSUE #7194: Avoid effect instrumentation for pure non-control-impure functions
Right now we add effect instrumentation to all functions. This is unnecessary is the function does not have unhandled effects.

This would normally not be true since a pure function can use effects inside a try block, but since we lift the try body into a new function, this is not a problem

### PR DESCRIPTION
**Pull request by JonathanStarup**
Fixes #7013
Fixes  #7194

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/ast/Purity.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Purity.scala:108]: Comment

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:43]: It just seems ad-hoc.

**Comment by magnus-madsen** [comment]
Ill let @mlutze have a look too, since he is working on adjacent things.

**Comment by magnus-madsen** [comment]
Is it right that you are already using this information in the backend?

File: main/src/ca/uwaterloo/flix/language/ast/Purity.scala
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/ast/Purity.scala:110]: I don't think so. You could maybe make it less characters but it wouldn't be easier to read in my opinion.

File: main/src/ca/uwaterloo/flix/language/ast/Purity.scala
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/ast/Purity.scala:68]: I'm not sure which line youre refering to since it shows me four lines, but controlpure has to be above impure. I've moved Pure down to the bottom though

**Comment by JonathanStarup** [comment]
> Is it right that you are already using this information in the backend?

Yes, I can remove that again if you want (1 line change) but i just wanted to make sure it worked

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:43]: hmm yeah. I do also like that its specific though, you don't have to wonder about how `root` is used.

**Review by mlutze — APPROVED** [review]

File: main/src/ca/uwaterloo/flix/language/ast/Purity.scala
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Purity.scala:67]: Actually, is there a reason not to spell out all cases? There are only 9 right?

**Comment by magnus-madsen** [comment]
> > Is it right that you are already using this information in the backend?
> 
> Yes, I can remove that again if you want (1 line change) but i just wanted to make sure it worked

If the tests pass its OK.

File: main/src/ca/uwaterloo/flix/language/ast/Purity.scala
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/ast/Purity.scala:67]: 5 more places to typo perhaps, but I can do that. I could also make a `Purity => Int` function and use that

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:43]: @mlutze ?

**Comment by magnus-madsen** [comment]
Don't merge until I do one more pass. My initial pass was just that: initial.

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:43]: Given that the result is going to be finite or universe-minus-finite, could we just represent that symbolically and avoid dealing with the universe altogether?

Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:43]: I like it 😄 discussed with magnus somewhere else, but I does add almost 100 lines and a fair bit of places to mess up logically with 9 case union/intersection

**Comment by JonathanStarup** [comment]
I've put the refactoring part into this PR #7196 which should make this PR way simpler when rebased

**Comment by magnus-madsen** [comment]
RFR?

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala]: found out its only two case classes, so not as bad

**Comment by magnus-madsen** [comment]
@mlutze What do you think?

File: main/src/ca/uwaterloo/flix/language/ast/Purity.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Purity.scala:116]: I would worry that this check is not always complete.

Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/ast/Purity.scala:116]: Yeah, but it is wrong to the safe side. Additionally - given the conjecture that no actual effects will be infinite (even with an infinite effect domain) - it is accurate since evaluateFormula will never return `Compl`

**Comment by magnus-madsen** [comment]
Having slept on it, I would prefer to drop InfSet. We can bring it back later if this is something we want to broadly support. Right now I think it is just unnecessary indirection and abstraction.

File: main/src/ca/uwaterloo/flix/language/phase/Inliner.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:299]: @JonathanStarup I am not sure where this function is used, but this seems dangerous to me. Would this not end up removing expressions with control-effects?

Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/Inliner.scala]: Its just the wording thats misleading. I've changed it to `isDeadAndNotPure` now.

File: main/src/ca/uwaterloo/flix/language/phase/Inliner.scala
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:200]: Yea dont know. I assumed the writer wanted the branches to be correct no matter the order, but it could just check for dead

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:200]: Dead should be what we want?

Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/Inliner.scala]: I've rewritten it now. Is this clearer?

File: main/src/ca/uwaterloo/flix/language/phase/Inliner.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:194]: Why not just isPure and flip the conditions?

Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/Inliner.scala]: fixed

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
Thanks!

**Pull request merged**
